### PR TITLE
Fix TensorImpl use count assert for _thnn_fused_lstm_cell_backward

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76316

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -5833,9 +5833,13 @@
...
 
-- func: _thnn_fused_lstm_cell_backward(Tensor? grad_hy, Tensor? grad_cy, Tensor cx, Tensor cy, Tensor workspace, bool has_bias) -> (Tensor, Tensor, Tensor, Tensor, Tensor)
+# NB: The composite version of this function below is a simple wrapper that duplicates some of the outputs
+#     It is necessary to avoid triggering TensorImpl use count checks in debug mode
+- func: _thnn_fused_lstm_cell_backward_impl(Tensor? grad_hy, Tensor? grad_cy, Tensor cx, Tensor cy, Tensor workspace, bool has_bias) -> (Tensor, Tensor, Tensor)
   dispatch:
-    CUDA: _thnn_fused_lstm_cell_backward_cuda
+    CUDA: _thnn_fused_lstm_cell_backward_cuda_impl
+
+- func: _thnn_fused_lstm_cell_backward(Tensor? grad_hy, Tensor? grad_cy, Tensor cx, Tensor cy, Tensor workspace, bool has_bias) -> (Tensor, Tensor, Tensor, Tensor, Tensor)
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:5842]: nit: we might want to have a comment here mentioning that this function is NOT differentiable. Also should we have a this in the impl `TORCH_INTERNAL_ASSERT(!GradMode::enabled());` ?

**Review by albanD — APPROVED** [review]
SGTM

Maybe an extra assert would be nice

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (soulitzer) [reply aten/src/ATen/native/native_functions.yaml:5842]: Hmm, don't we already have the autograd not implemented kernel for the case when someone actually tries to double backward on this node? In theory someone could have `create_graph=True` but only `.grad` part of the graph.

Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:5843]: But then this wouldn't be called?

**Review by albanD — APPROVED** [review]

**Pull request closed without merging**